File: src/DashSpec.Core/Parsing/DataSourceParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class DataSourceParser
{
    public static DataSourceDefinition Parse(TokenReader reader, string? specDirectory = null)
    {
        if (reader.TryKeyword("view"))
        {
            var name = reader.ReadQualifiedName();
            SqlReadOnlyValidator.ValidateViewReference(name);
            return new DataSourceDefinition(DataSourceKind.View, name);
        }

        if (reader.TryKeyword("sql"))
        {
            if (reader.IsAt(TokenKind.LBrace))
            {
                return ParseSqlBlock(reader, specDirectory);
            }

            return ParseSqlInline(reader, specDirectory);
        }

        throw reader.Unexpected("view or sql");
    }

    private static DataSourceDefinition ParseSqlInline(TokenReader reader, string? specDirectory)
    {
        if (reader.TryKeyword("query"))
        {
            var body = ReadSqlQueryText(reader);
            SqlReadOnlyValidator.ValidateSqlBody(body);
            return new DataSourceDefinition(DataSourceKind.Sql, body, DataSourceSqlCarrier.Query);
        }

        if (reader.TryKeyword("file"))
        {
            var path = ReadSqlFileReference(reader);
            ValidateSqlFileExists(path, specDirectory);
            return new DataSourceDefinition(DataSourceKind.Sql, path, DataSourceSqlCarrier.File);
        }

        throw new DashSpecParseException(
            "datasource sql requires 'query' or 'file' (e.g. datasource sql query \"SELECT …\" or datasource sql file \"sql/x.sql\").");
    }

    private static DataSourceDefinition ParseSqlBlock(TokenReader reader, string? specDirectory)
    {
        reader.Expect(TokenKind.LBrace);
        reader.SkipNewlines();

        DataSourceDefinition? parsed = null;

        while (!reader.IsAt(TokenKind.RBrace) && !reader.IsEof)
        {
            reader.SkipNewlines();
            if (reader.IsAt(TokenKind.RBrace))
            {
                break;
            }

            if (!reader.TryKeyword("from"))
            {
                throw reader.Unexpected("from");
            }

            if (reader.TryKeyword("query"))
            {
                var body = ReadSqlQueryText(reader);
                SqlReadOnlyValidator.ValidateSqlBody(body);
                parsed = new DataSourceDefinition(DataSourceKind.Sql, body, DataSourceSqlCarrier.Query);
            }
            else if (reader.TryKeyword("file"))
            {
                var path = ReadSqlFileReference(reader);
                ValidateSqlFileExists(path, specDirectory);
                parsed = new DataSourceDefinition(DataSourceKind.Sql, path, DataSourceSqlCarrier.File);
            }
            else
            {
                throw reader.Unexpected("query or file after from");
            }

            reader.SkipNewlines();
        }

        reader.Expect(TokenKind.RBrace);

        if (parsed is null)
        {
            throw new DashSpecParseException("datasource sql { } requires from query or from file.");
        }

        return parsed;
    }

    private static string ReadSqlQueryText(TokenReader reader)
    {
        reader.SkipNewlines();
        return reader.CurrentKind switch
        {
            TokenKind.String => reader.ReadString(),
            TokenKind.Raw => UnwrapRawSql(reader.ReadRawBlock()),
            _ => throw reader.Unexpected("SQL query string or [[ … ]] block"),
        };
    }

    private static string ReadSqlFileReference(TokenReader reader)
    {
        reader.SkipNewlines();
        if (reader.CurrentKind is not TokenKind.String)
        {
            throw reader.Unexpected("file path string");
        }

        var path = reader.ReadString();
        if (string.IsNullOrWhiteSpace(path))
        {
            throw new DashSpecParseException("datasource sql file path must not be empty.");
        }

        return path.Replace('\\', '/');
    }

    private static void ValidateSqlFileExists(string relativePath, string? specDirectory)
    {
        if (string.IsNullOrWhiteSpace(specDirectory))
        {
            return;
        }

        var path = Path.GetFullPath(Path.Combine(specDirectory, relativePath));
        if (!File.Exists(path))
        {
            throw new FileNotFoundException(
                $"datasource sql file not found: '{relativePath}' (resolved: {path}).",
                path);
        }
    }

    private static string UnwrapRawSql(string raw)
    {
        var trimmed = raw.Trim();
        if (trimmed.StartsWith("[[", StringComparison.Ordinal) &&
            trimmed.EndsWith("]]", StringComparison.Ordinal))
        {
            return trimmed[2..^2].Trim();
        }

        return trimmed;
    }
}

File: src/DashSpec.Modeling.Parse/DataSource/DataSourceParser.fs
namespace DashSpec.Modeling.Parse.DataSource

open System
open System.IO
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module DataSourceParser =

    let private unwrapRawSql (raw: string) =
        let trimmed = raw.Trim()
        if trimmed.StartsWith("[[", StringComparison.Ordinal) && trimmed.EndsWith("]]", StringComparison.Ordinal) then
            trimmed.Substring(2, trimmed.Length - 4).Trim()
        else
            trimmed

    let private readSqlQueryText (reader: TokenReader) =
        reader.SkipNewlines()
        match reader.CurrentKind with
        | TokenKind.String -> reader.ReadString()
        | TokenKind.Raw -> unwrapRawSql (reader.ReadRawBlock())
        | _ -> raise (reader.Unexpected "SQL query string or [[ … ]] block")

    let private readSqlFileReference (reader: TokenReader) =
        reader.SkipNewlines()
        if reader.CurrentKind <> TokenKind.String then
            raise (reader.Unexpected "file path string")
        let path = reader.ReadString()
        if String.IsNullOrWhiteSpace path then
            raise (DashSpecParseException("datasource sql file path must not be empty."))
        path.Replace('\\', '/')

    let private validateSqlFileExists (relativePath: string) (specDirectory: string option) =
        match specDirectory with
        | None | Some (null | "") -> ()
        | Some dir ->
            let path = Path.GetFullPath(Path.Combine(dir, relativePath))
            if not (File.Exists path) then
                raise (FileNotFoundException($"datasource sql file not found: '{relativePath}' (resolved: {path}).", path))

    let private parseSqlInline (reader: TokenReader) (specDirectory: string option) =
        if reader.TryKeyword "query" then
            let body = readSqlQueryText reader
            SqlReadOnlyValidator.validateSqlBody body
            { Kind = DataSourceKind.Sql; Value = body; SqlCarrier = Some DataSourceSqlCarrier.Query }
        elif reader.TryKeyword "file" then
            let path = readSqlFileReference reader
            validateSqlFileExists path specDirectory
            { Kind = DataSourceKind.Sql; Value = path; SqlCarrier = Some DataSourceSqlCarrier.File }
        else
            raise (DashSpecParseException("datasource sql requires 'query' or 'file' (e.g. datasource sql query \"SELECT …\" or datasource sql file \"sql/x.sql\")."))

    let private parseSqlBlock (reader: TokenReader) (specDirectory: string option) =
        reader.Expect TokenKind.LBrace
        reader.SkipNewlines()
        let mutable parsed: DataSourceDefinition option = None

        while not (reader.IsAt TokenKind.RBrace) && not reader.IsEof do
            reader.SkipNewlines()
            if reader.IsAt TokenKind.RBrace then ()
            elif not (reader.TryKeyword "from") then
                raise (reader.Unexpected "from")
            elif reader.TryKeyword "query" then
                let body = readSqlQueryText reader
                SqlReadOnlyValidator.validateSqlBody body
                parsed <- Some { Kind = DataSourceKind.Sql; Value = body; SqlCarrier = Some DataSourceSqlCarrier.Query }
            elif reader.TryKeyword "file" then
                let path = readSqlFileReference reader
                validateSqlFileExists path specDirectory
                parsed <- Some { Kind = DataSourceKind.Sql; Value = path; SqlCarrier = Some DataSourceSqlCarrier.File }
            else
                raise (reader.Unexpected "query or file after from")
            reader.SkipNewlines()

        reader.Expect TokenKind.RBrace

        match parsed with
        | Some value -> value
        | None -> raise (DashSpecParseException("datasource sql { } requires from query or from file."))

    let parse (reader: TokenReader) (specDirectory: string option) =
        if reader.TryKeyword "view" then
            let name = reader.ReadQualifiedName()
            SqlReadOnlyValidator.validateViewReference name
            { Kind = DataSourceKind.View; Value = name; SqlCarrier = None }
        elif reader.TryKeyword "sql" then
            if reader.IsAt TokenKind.LBrace then parseSqlBlock reader specDirectory
            else parseSqlInline reader specDirectory
        else
            raise (reader.Unexpected "view or sql")
